File: Iteration 1/code/CSE3063GRP12/src/core/repositories/UserRepository.java
package core.repositories;

import java.util.ArrayList;

import com.google.gson.Gson;

import java.io.IOException;
import core.database.abstracts.DatabaseManager;
import core.enums.UserType;
import core.exceptions.UserNotFoundException;
import core.exceptions.WrongPasswordException;
import core.general_providers.AppConstant;
import core.general_providers.InstanceManager;
import core.general_providers.SessionController;
import core.models.abstracts.User;
import core.models.concretes.Advisor;
import core.models.concretes.Student;

public class UserRepository {
    private DatabaseManager databaseManager;
    private String path;
    private String advisorPath;
    private String studentPath;

    public UserRepository() {
        path = System.getProperty("user.dir") + AppConstant.getInstance().getBasePath() + "/user/";
        advisorPath = path + "advisor/";
        studentPath = path + "student/";
        databaseManager = InstanceManager.getInstance().getDataBaseInstance();

    }

    public void loginCheck(String userName, String password)
            throws UserNotFoundException, WrongPasswordException {
        try {
            User user = null;
            if (getUserType(userName) == UserType.Student) {
                user = databaseManager.read(studentPath + "/" + userName + ".json", Student.class);
                if (user.getPassword().equals(password)) {
                    setCurrentUser(user);
                } else {
                    throw new WrongPasswordException();
                }

            } else if (getUserType(userName) == UserType.Advisor) {
                user = databaseManager.read(advisorPath + "/" + userName + ".json", Advisor.class);
                if (user.getPassword().equals(password)) {
                    setCurrentUser(user);

                } else {
                    throw new WrongPasswordException();
                }

            }
            // TODO: PROBABLY IT WONT WORK
            if (user == null) {
                throw new UserNotFoundException();
            }

        } catch (Exception e) {
            e.printStackTrace();
            throw new UserNotFoundException();
        }
    }

    public void setCurrentUser(User user) throws IOException {
        SessionController.getInstance().setCurrentUser(user);
    }

    public ArrayList<Student> getStudentsByAdvisor(Advisor advisor) {
        try {
            ArrayList<Student> students = new ArrayList<Student>();
            for (String ids : advisor.getListOfStudentIds()) {
                Student student = databaseManager.read(studentPath + "/" + ids + ".json", Student.class);
                students.add(student);
            }
            return students;
        } catch (Exception e) {
            e.printStackTrace();
        }

        return null;
    }

    public User getUser(String userName) throws UserNotFoundException {
        User user = null;
        try {

            if (getUserType(userName) == UserType.Student) {
                user = databaseManager.read(studentPath + "/" + userName + ".json", Student.class);
            } else if (getUserType(userName) == UserType.Advisor) {
                user = databaseManager.read(advisorPath + "/" + userName + ".json", Advisor.class);
            }
            // TODO: PROBABLY IT WONT WORK
            if (user == null) {
                throw new UserNotFoundException();
            }

        } catch (Exception e) {
            e.printStackTrace();
            throw new UserNotFoundException();
        }
        return user;

    }

    private UserType getUserType(String userName) {
        char firstChar = userName.charAt(0);
        if (firstChar == 'o') {
            return UserType.Student;
        } else if (firstChar == 'a') {
            return UserType.Advisor;
        }
        return null;
    }

}

File: Iteration3/code/CSE3063GRP12/src/core/repositories/UserRepository.py
import os
from core.models.concretes.Advisor import Advisor
from core.models.concretes.Student import Student
from core.database.concretes.JsonManager import JsonManager
from core.exceptions.UserNotFoundException import UserNotFoundException
from core.exceptions.WrongPasswordException import WrongPasswordException
from core.general_providers.SessionController import SessionController
from core.general_providers.AppConstant import AppConstant
from core.general_providers.InstanceManager import InstanceManager
from core.enums.UserType import UserType

class UserRepository:
    def __init__(self):
        self.path = os.getcwd() + AppConstant.getInstance().get_base_path() + "user/"
        self.advisor_path = self.path + "advisor"
        self.student_path = self.path + "student"
        self.database_manager = InstanceManager().get_database_instance()

    def login_check(self, user_name, password):
        user = None
        user_type = self.get_user_type(user_name)
        if user_type == UserType.Student:
            user = self.database_manager.read(f"{self.student_path}/{user_name}.json", Student)
        elif user_type == UserType.Advisor:
            user = self.database_manager.read(f"{self.advisor_path}/{user_name}.json", Advisor)

       
        if user is None:
            raise UserNotFoundException()
        if user.password != password:
            raise WrongPasswordException()
        else:
            self.set_current_user(user)
      

    def set_current_user(self, user):
        SessionController.getInstance().set_current_user(user)


  
    def get_students_by_advisor(self, advisor):
        students = []
        for student_id in advisor.listOfStudentIds:
            student = self.database_manager.read(f"{self.student_path}/{student_id}.json", Student)
            students.append(student)
        return students


    def get_user(self, user_name):
        user = None
        user_type = self.get_user_type(user_name)
        if user_type == UserType.Student:
            user = self.database_manager.read(f"{self.student_path}/{user_name}.json", Student)
        elif user_type == UserType.Advisor:
            user = self.database_manager.read(f"{self.advisor_path}/{user_name}.json", Advisor)

        if user is None:
            raise UserNotFoundException()
     

        return user

    def get_user_type(self, user_name):
        if len(user_name) < 2:
            raise UserNotFoundException() 
        if user_name[0] == 'o':
            return UserType.Student
        elif user_name[0] == 'a':
            return UserType.Advisor
        else:
            raise UserNotFoundException()